### ISSUE TITLE
Support tap/tape

### ISSUE THREAD

**Issue by dylang** [wallabyjs/public#2]
https://github.com/isaacs/node-tap
https://github.com/substack/tape

**Comment by ArtemGovorov** [comment]
Thanks for the suggestion, will implement tap support in planned node.js runner.

**Comment by franleplant** [comment]
+1

**Comment by Pentiado** [comment]
Is it planned somewhere in near future? I would really like to switch to Tape, but, well without Wallaby there is no way I would decide to do it. Wallaby is too cool.

**Comment by ArtemGovorov** [comment]
@Pentiado I'm planning to have a look into supporting tape in a few weeks.

**Comment by NodeGuy** [comment]
:+1: 

Here's a strongly opinionated article about why it's great: [Why I use Tape Instead of Mocha & So Should You](https://medium.com/javascript-scene/why-i-use-tape-instead-of-mocha-so-should-you-6aa105d8eaf4)

**Comment by Nembus** [comment]
I too would love to see Tape support.

**Comment by willhoag** [comment]
Yep, I'm a Tape user as well and would love support for it. Thanks!

**Comment by geraldyeo** [comment]
Supporting Tape will be awesome!

**Comment by TYRONEMICHAEL** [comment]
+1. Very interested as well.

**Comment by WRidder** [comment]
+1 as well

**Comment by davegomez** [comment]
Any progress on this or should I install Mocha?

**Comment by ArtemGovorov** [comment]
@davegomez we have made some progress, but no ETA yet as there're still a few things to do.

**Comment by ArtemGovorov** [comment]
A quick poll: are you using tape for node or for browser tests?

**Comment by davegomez** [comment]
I'm currently using it for Node but I would love to use it for browsers too (but the config with PhantomJS and Karma is a pain in the ass).

**Comment by Nembus** [comment]
Browser mostly here.

**Comment by TYRONEMICHAEL** [comment]
Node.

**Comment by willhoag** [comment]
Both

**Comment by WRidder** [comment]
Both as well.

**Comment by geraldyeo** [comment]
Both

**Comment by baldore** [comment]
Both.

**Comment by iclanzan** [comment]
Both

**Comment by WRidder** [comment]
I was wondering whether there are any updates w.r.t. tape support?

Writing tests is much less fun without Wallaby ;)

**Comment by ArtemGovorov** [comment]
@WRidder We are close to finish tape for node.js tests support, still a few things to do, but it is getting there.

**Comment by WRidder** [comment]
Awesome! Looking forward to it. Your support is impeccable :)

**Comment by GuidoS** [comment]
+1

**Comment by bhishp** [comment]
+1

**Comment by ArtemGovorov** [comment]
If anyone's willing to have a look, this is currently the blocking issue for Tape support in wallaby.js - https://github.com/substack/tape/issues/222

**Comment by QuantumInformation** [comment]
We are using tape for browser.

**Comment by gnavvy** [comment]
any progress on the tape support?

**Comment by ArtemGovorov** [comment]
There's a highly experimental support available for node.js tests with tape in wallaby.js:

```javascript
module.exports = function () {
  return {
    ...
    testFramework: 'tape',
    env: {
      type: 'node',
      runner: 'node'
    }
  };
};
```

**Comment by ArtemGovorov** [comment]
This is still not resolved in tape: https://github.com/substack/tape/issues/222, if there're any issues with your tape tests running with wallaby, check if that's not that issue.

**Comment by pietvanzoen** [comment]
Thanks for adding support! Even if it is experimental, it's super handy.

**Comment by ArtemGovorov** [comment]
@pietvanzoen Awesome, thanks for checking! We have had few private beta users testing it as well, worked for everyone quite well, no known issues except the one I referenced.

**Comment by rainydio** [comment]
@ArtemGovorov can you please improve tape runner slightly? When using [tape-promise](https://github.com/jprichardson/tape-promise) stack traces aren't shown. As far as I can see `tape` has special field `actual.stack` which it actually [checks and prints](https://github.com/substack/tape/blob/master/lib/results.js#L160) during report.

If you can check `actual.stack` before `error.stack` this will bring stack traces. Also in addition to `fail` operator you'll have to check `error` operator before adding `Actual: ... Expected: ...` to the error message.

**Comment by rainydio** [comment]
Well, after another update I'm still patching the runner. Here is the diff, this version is somewhat cleaner because it actually checks that `operator` is `"error"`. Can you implement it please?

```diff
if (!i.skipped)
    for (var s = 0, r = t._log.length; r > s; s++) {
        var a = t._log[s];
-       i.success = i.success && a.ok;
-       if (!a.ok && a.operator === "error") {
-           var o = {
-             message: "Error: " + a.actual.message,
-             stack: a.actual.stack,
-             showDiff: false
-           };
-           tracer.setAssertionData(a, o);
-           i.log.push(o);
-       }
-       else if (!a.ok && a.error) {
+       if (i.success = i.success && a.ok, !a.ok && a.error) {
            var o = {
                message: a.error.message,
                stack: a.error.stack
```

**Comment by ArtemGovorov** [comment]
@rainydio Sorry about this one, your previous comment Github notification was lost in my inbox. Had pushed and published the change in core v1.0.301.

**Comment by ahmadnassri** [comment]
holding off / waiting anxiously getting a license till `tap` support is added :pray:

**Comment by Cyberlane** [comment]
@ahmadnassri I've been using it with `tape` for the past week, it seems to work just fine!

**Comment by mbcooper** [comment]
Is `tape` support coming soon?  I'm planning on using it, but can't lose wallaby.

**Comment by ArtemGovorov** [comment]
@mbcooper Wallaby has experimental support for `tape` for node, so you may try if it's working for you:

```javascript
module.exports = function () {
  return {
    ...
    testFramework: 'tape',
    env: {
      type: 'node',
      runner: 'node'
    }
  };
};
```

**Comment by ahmadnassri** [comment]
@Cyberlane `tape` vs `tap` ;)

**Comment by mbcooper** [comment]
Help. Trying it out. I am using React, and I'm having a challenge with imports of images

**Comment by ArtemGovorov** [comment]
@mbcooper Please create a separate issue and include your wallaby config (ideally a sample repo where the issue could be reproduced for the fastest resolution).

**Comment by pke** [comment]
@ArtemGovorov Holding back buying a license until `tap` is supported. Any chance to get this in?

**Comment by medikoo** [comment]
@ArtemGovorov can you explain in more detail on why https://github.com/substack/tape/issues/222 is a blocker? It appears that there's no clean fix for that possible (either native API's have to be hacked, or design of Tape slightly changed)

**Comment by ArtemGovorov** [comment]
@medikoo It's not a blocker as is, just may lead to some unpredictable results when using tape with wallaby, but you [may use it](https://github.com/wallabyjs/public/issues/2#issuecomment-281219306). It's just we can't fix that specific issue on our side, so the support is marked as experimental.

**Comment by medikoo** [comment]
@ArtemGovorov yes I'm using it, still I wondered what's the _specific issue_ on your side, that prevents announcing tape support as official

**Comment by ArtemGovorov** [comment]
@medikoo Nothing on our side, just the mentioned issue in tape that prevents marking the support to be official.

**Comment by ArtemGovorov** [comment]
@medikoo Unlike tape, which runs tests and stops the process, wallaby runs tests continuously and is able to run individual tests on test changes. This behaviour from tape may in certain cases break things in wallaby (mostly in the area test results reporting). If you [set the `workers: {recycle: true}` config property](https://wallabyjs.com/docs/config/workers.html) then wallaby will work more like tape does, and the consequences of the issue will be minimised.

**Comment by medikoo** [comment]
> If you set the workers: {recycle: true} config property then wallaby will work more like tape does

I very quickly found out that I need this setting anyway, and it's not `tape` related. e.g. it's needed for [sinon](http://sinonjs.org/) stubs, The `sinon.stub(someObj, "method")` will crash with no `{ recycle: true }`.

It means that probably having `{ recycle: false }` as default is maybe not good idea in general

**Comment by davegomez** [comment]
> It means that probably having { recycle: false } as default is maybe not that good idea in general

I strongly disagree on that.

**Comment by ArtemGovorov** [comment]
@medikoo The `recycle: true` is only required if you prefer not to clean up after your tests. Using `sinon.stub` without cleaning up may affect your tests not only when running in wallaby, but for example when accessing the same (stubbed) API in different tests (or test files). [Sinon provides a way to undo all stubs](http://stackoverflow.com/questions/11552991/cleaning-up-sinon-stubs-easily).

> It means that probably having `{ recycle: false }` as default is maybe not that good idea in general

The setting's default was working for the majority of our users since day 1 (for over 2 years), and so far we've only seen very few cases when a proper test cleanup was not possible or was not desirable (in which case `{ recycle: true }` helped).

Also re-using test runner node processes allows wallaby to run your tests faster than when recycling node processes.

**Comment by ahmadnassri** [comment]
can we please split `tap` and `tape` into separate issues? seems like this thread is mainly focused on `tape` _(which is totally okay :+1:)_ but I will have to mute and loose ability to track the `tap` side of the conversation 😊

**Comment by ahmadnassri** [comment]
@ArtemGovorov thank you 👏

**Comment by medikoo** [comment]
> @medikoo The recycle: true is normally only required if you prefer not to clean up after your tests. 

Indeed with `overridenMethod.restore()` there's no need for `recycle: false`. Thanks for clarification.

**Comment by eggyal** [comment]
Why no support (even experimental) for tape in browser?  Are there upstream issues blocking that too?
